src/Streamly/Coreutils/Cp.hs

```diff
@@ -0,0 +1,281 @@
+    -> FilePath
+    -> FilePath
+    -> IO Bool
+copy opt src dst = do
+    existDst <- doesFileExist dst
+    existSrc <- doesFileExist src
+    if not existSrc
+    then do
+        putStrLn $ "Source file " ++ src ++ " does not exist"
+        return False
+    else if existDst && not (overwriteExisting opt)
+    then do
+        putStrLn $ "overwriteExisting : False " ++ dst ++ " already exists"
+        return False
+    else if verbose opt == True
+    then do
+        File.fromChunks dst $ File.toChunksWithBufferOf (256*1024) src
+        putStrLn $ src ++ " -> " ++ dst
+        return True
+    else do
+        File.fromChunks dst $ File.toChunksWithBufferOf (256*1024) src
+        return True
+
+
+-- | Append two paths - first one should be a directory
+-- Assumes second path is relative and without a leading '/'
+--
+-- @since 0.1.0.0
+{-# INLINE append #-}
+append
+    :: FilePath
+    -> FilePath
+    -> FilePath
+append dir file =
+    let end = last dir
+        osName = os
+        pathSep = if osName == "windows" then '\\' else '/'
+     in
+         if dir == ""
+         then file
+         else if end == pathSep
+         then dir ++ file
+         else dir ++ [pathSep] ++ file
+
+
+-- | Removes a trailing slash from a file path
+--
+-- @since 0.1.0.0
+{-# INLINE removeTrailingSlash #-}
+removeTrailingSlash
+    :: FilePath
+    -> FilePath
+removeTrailingSlash "" = ""
+removeTrailingSlash path =
+    let end = last path
+        osName = os
+        pathSep = if osName == "windows" then '\\' else '/'
+     in
+         if end == pathSep
+         then init path
+         else path
+
+
+-- | Lists the contents (both files and directories) of the directory
+-- relative to the argument directory.
+-- @Left path@ represents that @path@ is a directory while @Right path@
+-- represents that it is a file.
+--
+-- @since 0.1.0.0
+traverseDir
+    :: IsStream t
+    => FilePath
+    -> t IO (Either FilePath FilePath)
+traverseDir baseDir = do
+    IP.concatMapTreeWith ahead listContents $ S.yield $ Left ""
+
+    where
+
+    listContents :: IsStream t => FilePath -> t IO (Either FilePath FilePath)
+    listContents dir = do
+        S.fromListM $ map (identify dir) $ contents dir
+
+    identify :: FilePath -> FilePath -> IO (Either FilePath FilePath)
+    identify dir path = do
+       let actualPath = append baseDir $ append dir path
+       isFile <- doesFileExist actualPath
+       if isFile == True
+       then
+           return $ Right $ append dir path     -- RELATIVE File
+       else
+           return $ Left $ append dir path      -- RELATIVE Dir
+
+    contents :: FilePath -> [FilePath]
+    contents dir = unsafePerformIO $ listDirectory $ append baseDir dir
+
+
+
+-- | Recursively copies the contents of the source directory
+-- to the destination directory.
+-- Arguments are of type @FilePath@.
+--
+-- @since 0.1.0.0
+copyDirToDir
+    :: IsStream t
+    => CpOptions
+    -> FilePath
+    -> FilePath
+    -> t IO ()
```

Comment:
`copyDirToDir :: CpOptions -> Path Abs Dir -> Path Abs Dir -> IO ()`?
Why `t IO ()`?

Reply:
I did not drain the stream so that user can use the appropriate combinator like `asyncly`, `wAsyncly` to achieve concurrency.